pkg/resources/mutation/cluster/provisioningcluster.go

```diff
@@ -53,10 +97,257 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
+				cluster = SetKubeAPIServerArg(args, cluster)
+			}
+		} else {
+			// Now, handle the case of PSACT being set when creating or updating the cluster
+			template, err := m.psact.Get(templateName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return admission.ResponseBadRequest(err.Error()), nil
+				}
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to get psact: %w", err)
+			}
+			fileContent, err := psa.GenerateAdmissionConfigFile(template, cluster.Spec.KubernetesVersion)
+			if err != nil {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to generate admission configuration from PSACT [%s]: %w", templateName, err)
+			}
+			anno := map[string]string{
+				SecretAnnotation: cluster.Name,
+			}
+			data := map[string][]byte{
+				SecretKey: fileContent,
+			}
+			_, err = m.createOrUpdateSecret(cluster.Namespace, secretName, data, anno)
+			if err != nil {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to create or update the secret for the admission configuration file: %w", err)
+			}
+			// set relevant fields in the cluster
+			cluster = AddMachineSelectorFile(MachineSelectorFileForPSA(secretName, mountPath), cluster)
+			args := GetKubeAPIServerArg(cluster)
+			args[KubeAPIAdmissionConfigOption] = mountPath
+			cluster = SetKubeAPIServerArg(args, cluster)
+		}
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+// createOrUpdateSecret creates or updates a secret based on the provided information.
+func (m *ProvisioningClusterMutator) createOrUpdateSecret(namespace, name string, data map[string][]byte, annotations map[string]string) (*corev1.Secret, error) {
+	var existing *corev1.Secret
+	var err error
+	if namespace == "" || name == "" {
+		return existing, fmt.Errorf("namespace and name are required")
+	}
+	existing, err = m.secret.Cache().Get(namespace, name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: data,
+		Type: corev1.SecretTypeOpaque,
+	}
+	if annotations != nil {
+		secret.Annotations = annotations
+	}
+	if existing == nil {
+		return m.secret.Create(secret)
+	}
+	if !reflect.DeepEqual(existing.Data, secret.Data) {
+		existing.Data = data
+		return m.secret.Update(existing)
+	}
+	return secret, nil
+}
+
+// GetKubeAPIServerArg returns a map representation of the value of kube-apiserver-arg from the cluster's MachineGlobalConfig.
+// An empty map is returned if kube-apiserver-arg is not set in the cluster.
+func GetKubeAPIServerArg(cluster *v1.Cluster) map[string]string {
+	if cluster.Spec.RKEConfig.MachineGlobalConfig.Data != nil {
+		return toMap(cluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"])
+	}
+	return map[string]string{}
+}
+
+func toMap(input interface{}) map[string]string {
+	args := map[string]string{}
+	parsed := convert.ToInterfaceSlice(input)
+	if parsed != nil {
```

Comment:
We do not need a nil check since range can handle a nil slice.
```suggestion
```

Reply:
fixed.